### PR TITLE
fixes https://github.com/qooxdoo/qooxdoo-compiler/issues/692 & adds s

### PR DESCRIPTION
**Pull request by johnspackman**
…upport for worker class initialisation
update schemas
reverse the workerClass

Replaces https://github.com/qooxdoo/qooxdoo-compiler/pull/775

### CODE REVIEW

**Review by hkollmann — APPROVED** [review]

**Comment by johnspackman** [comment]
Please don't merge this; I have a few other fixes that are coming together and which would be much easier to merge as a single unit, and which are based on the code in this PR

File: Qooxdoo Compiler.iml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cboulanger) [inline Qooxdoo Compiler.iml:1]: This file should probably not be committed?

File: README.md [hunk 2]
```diff
@@ -54,7 +54,7 @@ nvm use 8 # or 10
 ## Test drive
 
 For more detailed information about installation and use of the compiler, refer
-to the [documentation](https://www.qooxdoo.org/docs/#/cli/commands). 
+to the [documentation](https://qooxdoo.org/documentation/#/development/compiler/?id=installation-getting-started).
```
Comment (cboulanger) [inline README.md:57]: Mabe without the querystring part (since the heading might change)?

**Review by cboulanger — APPROVED** [review]

**Comment by johnspackman** [comment]
This PR is now a lot larger, but includes various PRs that were building up ready to go; in summary, they are:

API events are called asynchronously;
Privates are mangled properly in strings (fixes https://github.com/qooxdoo/qooxdoo-compiler/issues/715);
Add `qx.compiler` env setting;
Add support for mixin methods to call `this.base` (see https://github.com/qooxdoo/qooxdoo/pull/10073);
Fixes a bug with file watch not working (on mac);
Fixes a bug with uglify suppressing all comments, including those which need to remain (eg for hints);
Fixes a bug in dependon/requiredby scripts;
Fixes a bug where polyfills loaded incorrectly (fixes https://github.com/qooxdoo/qooxdoo-compiler/issues/692)

@hkollmann as discussed, this will break the API viewer; the problem is that the "checkEnvironment" event is fired before the .json meta data files are written, and as the events are now fired asynchronously and with `await`, the promise in API Viewer's compile.js will be waited for, so make this more obvious.

If you have a look at the events in `qx.tool.cli.commands.Compile`, you'll see that they are now fired with different data, one of which is `appMeta` which is the instance of `qx.tool.compiler.targets.meta.ApplicationMeta` - this is the "stream-in-progress", the output of the application before it is serialised to disk, and you can use it to manipulate the application before it is written.  

For example, something like `appMeta.addPreBootCode` allows you to add code to the `PreBootCode` macro referred to at https://github.com/qooxdoo/qooxdoo-compiler/blob/master/source/resource/qx/tool/cli/templates/loader/loader-browser.tmpl.js#L496

That's probably the simplest way to inject code, I think it will do what you need - let me know if you need anything else, it should just be a matter of pointing you to the right API call now (bugs notwithstanding 😀 )

**Comment by johnspackman** [comment]
Does anyone know what the hell is wrong with the tests?  According to the tests console output, the "known good" compiler is using an old version of the framework (hence all the warnings like "warning: Environment check 'html.webworker' may be indeterminable,..."), but the tests also show that ` npm list @qooxdoo/framework` returns:

```
@qooxdoo/compiler@1.0.0-beta /home/runner/work/qooxdoo-compiler/qooxdoo-compiler
├─┬ @qooxdoo/compiler@1.0.0-beta-20201203-1458
│ └── @qooxdoo/framework@6.0.0-beta-20201203-1458 
└── @qooxdoo/framework@6.0.0-beta-20210105-1518 
```

I have checked out a fresh copy of my repo into a new directory, `git clean -dfX`, and run the `npm ci ;  npm list @qooxdoo/framework` and I get:

```
@qooxdoo/compiler@1.0.0-beta /Users/john/dev/Qooxdoo/tmp/qooxdoo-compiler
├─┬ @qooxdoo/compiler@1.0.0-beta-20201203-1458
│ └── @qooxdoo/framework@6.0.0-beta-20201203-1458
└── @qooxdoo/framework@6.0.0-beta-20210105-1518
```

So in theory, the test will use exactly the same "known good" version of the compiler and the same version of the framework as my new copy.  Except it doesn't.  The output on test is broken, and the crash that happens later on is because the compiler is the wrong version or something along those lines.  More to the point, `npm test` works just fine on mine, bootstrapping without loads of warnings and produces the new compiler.

**Comment by johnspackman** [comment]
I've SSH'd into github action and what is going on here:

```
runner@fv-az151-212:~/work/qooxdoo-compiler/qooxdoo-compiler$ git clean -dfX
Removing node_modules/
runner@fv-az151-212:~/work/qooxdoo-compiler/qooxdoo-compiler$ npx qx --version
npx: installed 2 in 1.431s
command not found: qx
runner@fv-az151-212:~/work/qooxdoo-compiler/qooxdoo-compiler$ grep 'qooxdoo/compiler' *.json
Manifest.json:    "@qooxdoo/compiler": "^1.0.0-beta"
npm-shrinkwrap.json:  "name": "@qooxdoo/compiler",
npm-shrinkwrap.json:    "@qooxdoo/compiler": {
npm-shrinkwrap.json:      "resolved": "https://registry.npmjs.org/@qooxdoo/compiler/-/compiler-1.0.0-beta-20201130-0808.tgz",
package.json:  "name": "@qooxdoo/compiler",
package.json:    "@qooxdoo/compiler": "1.0.0-beta-20201130-0808",
runner@fv-az151-212:~/work/qooxdoo-compiler/qooxdoo-compiler$ npm ci
[ ... snip ... ]
runner@fv-az151-212:~/work/qooxdoo-compiler/qooxdoo-compiler$ npx qx --version
1.0.0-beta-20201203-1458
```

In `npm-shrinkwrap.json` it says:
```
    "@qooxdoo/compiler": {
      "version": "1.0.0-beta-20201130-0808",
      "resolved": "https://registry.npmjs.org/@qooxdoo/compiler/-/compiler-1.0.0-beta-20201130-0808.tgz",
      "integrity": "sha512-T1jNHgQMwn0nREceYlSwWXZT9S8LOeDN5UghswIKQT5MOaXhq/GaH3ExwdS/IddC7MfFkaeisZgDjmZ4XqJFlQ==",
      "dev": true,
      "requires": {
```

So as far as I can make out, I have edited the npm config to always install *exactly* 1.0.0-beta-20201130-0808 of the compiler, but it chooses 1.0.0-beta-20201203-1458 - which is wrong, does not work, and isn't even the latest version.

**Comment by johnspackman** [comment]
PS I found the answer to my earlier issue - my machine has a global config of `qx.library` set to my current framework repo, that's why it's not reproducable

**Comment by cboulanger** [comment]
Wow, this bug seems to be quite complicated to fix! Have you figured out why it would install this particular outdated version?

**Comment by johnspackman** [comment]
finally got this going; my lesson is not to rely on tests on my own machine, particularly when working on the compiler and the framework together ☹️ 

there's some work on the deployment too, released just now

**Comment by hkollmann** [comment]
Ready for merge?

**Comment by johnspackman** [comment]
yes - I forgot to remove the label, doh!

**Pull request merged**